dsgrid/project.py

```diff
@@ -220,9 +201,151 @@ def _build_filtered_record_ids_by_dimension_type(self, context: QueryContext):
+            context.set_record_ids_by_dimension_type(dimension_type, ids)
+
+    def _process_dataset(
+        self, context: QueryContext, cached_datasets_dir: Path, dataset_id: str
+    ) -> Path:
+        """Return a Path to the created DataFrame. Does not return a DataFrame object because
+        the SparkSession will be restarted.
+
+        """
+        logger.info("Start processing query for dataset_id=%s", dataset_id)
+        project_version = f"{context.model.project.project_id}__{context.model.project.version}"
+        model_hash, text = context.model.project.dataset.params.serialize()
+        hash_dir = cached_datasets_dir / project_version / model_hash
+        if not hash_dir.exists():
+            hash_dir.mkdir(parents=True)
+            model_file = hash_dir / "model.json"
+            model_file.write_text(text)
+        cached_dataset_path = hash_dir / (dataset_id + ".parquet")
+        metadata_file = Path(str(cached_dataset_path).replace(".parquet", ".json5"))
+        if try_read_dataframe(cached_dataset_path) is None:
+            # An alternative solution is to call custom_spark_conf instead.
+            # That changes some settings without restarting the SparkSession.
+            # Results were not as good with that solution.
+            # Observations on queries with comstock and resstock showed that Spark
+            # used many fewer executors on the second query. That was with a standalone
+            # cluster on Eagle with dynamic allocation enabled.
+            # We don't understand why that is the case. It may not be an issue with YARN as
+            # the cluster manager on AWS.
+            # Queries on standalone clusters will be easier to debug if we restart the session
+            # for each big job.
+            with restart_spark_with_custom_conf(
+                conf=context.model.project.get_spark_conf(dataset_id),
+                force=True,
+            ):
+                logger.info("Build project-mapped dataset %s", dataset_id)
+                # Call load_dataset instead of get_dataset because the latter won't be valid here
+                # after the SparkSession restart.
+                dataset = self.load_dataset(dataset_id)
+                with Timer(timer_stats_collector, "build_project_mapped_dataset"):
+                    df = dataset.make_project_dataframe_from_query(context, self._config)
+                    write_dataframe_and_auto_partition(df, cached_dataset_path)
+                    context.serialize_dataset_metadata_to_file(dataset.dataset_id, metadata_file)
+        else:
+            assert metadata_file.exists(), metadata_file
+            context.set_dataset_metadata_from_file(dataset_id, metadata_file)
+            logger.info("Use cached project-mapped dataset %s", dataset_id)
+
+        logger.info("Finished processing query for dataset_id=%s", dataset_id)
+        return cached_dataset_path
+
+    def _process_exponential_growth_dataset(
```

Comment:
Is this temp or are we expecting to create one of these for each new dataset construction type?

Reply:
I'm expecting that we will need a custom function for any type that cannot be directly handled with DatasetExpressionHandler.